internal/apex/wg-iface.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		logger.Errorf("failed to start the wireguard listener: %v\n", err)
+	}
+
+	return nil
+}
+
+// setupLinuxInterface TODO replace with netlink calls
+// this is called if this is the first run or if the local node
+// address got assigned a new address by the controller
+func (ax *Apex) setupLinuxInterface(logger *zap.SugaredLogger) {
+	// delete the wireguard ip link interface if it exists
+	if ifaceExists(logger, wgIface) {
+		_, err := RunCommand("ip", "link", "del", wgIface)
+		if err != nil {
+			logger.Debugf("failed to delete the ip link interface: %v\n", err)
+		}
+	}
+	// create the wireguard ip link interface
+	_, err := RunCommand("ip", "link", "add", wgIface, "type", "wireguard")
+	if err != nil {
+		logger.Errorf("failed to create the ip link interface: %v\n", err)
+	}
+	// start the wireguard listener on a well-known port if it is the hub-router as all
+	// nodes need to be able to reach this node for state distribution if hole punching.
+	if ax.hubRouter {
+		_, err = RunCommand("wg", "set", wgIface, "listen-port", strconv.Itoa(WgDefaultPort), "private-key", linuxPrivateKeyFile)
+		if err != nil {
+			logger.Errorf("failed to start the wireguard listener: %v\n", err)
+		}
+	} else {
+		// start the wireguard listener
+		_, err = RunCommand("wg", "set", wgIface, "listen-port", strconv.Itoa(ax.listenPort), "private-key", linuxPrivateKeyFile)
+		if err != nil {
+			logger.Errorf("failed to start the wireguard listener: %v\n", err)
+		}
+	}
+	// give the wg interface an address
+	_, err = RunCommand("ip", "address", "add", ax.wgLocalAddress, "dev", wgIface)
+	if err != nil {
+		logger.Debugf("failed to assign an address to the local linux interface, attempting to flush the iface: %v\n", err)
+		wgIP := getIPv4Iface(wgIface)
+		_, err = RunCommand("ip", "address", "del", wgIP.To4().String(), "dev", wgIface)
+		if err != nil {
+			logger.Errorf("failed to assign an address to the local linux interface: %v\n", err)
+		}
+		_, err = RunCommand("ip", "address", "add", ax.wgLocalAddress, "dev", wgIface)
+		if err != nil {
+			logger.Errorf("failed to assign an address to the local linux interface: %v\n", err)
+		}
+	}
+	// bring the wg0 interface up
+	_, err = RunCommand("ip", "link", "set", wgIface, "up")
+	if err != nil {
+		logger.Errorf("failed to bring up the wg interface: %v\n", err)
+	}
+}
+
+func setupWindowsIface(logger *zap.SugaredLogger, localAddress, privateKey string) error {
+	if err := buildWindowsWireguardIfaceConf(privateKey, localAddress); err != nil {
+		return fmt.Errorf("failed to create the windows wireguard wg0 interface file %v", err)
+	}
+
+	var wgOut string
+	var err error
+	if ifaceExists(logger, wgIface) {
+		wgOut, err = RunCommand("wireguard.exe", "/uninstalltunnelservice", wgIface)
+		if err != nil {
+			logger.Debugf("failed to down the wireguard interface (this is generally ok): %v\n", err)
+		}
+		time.Sleep(time.Second * 1)
```

Comment:
I was wondering why the sleep was here, but then I see the comment a few lines down.  It looks like it should be moved up?

The comment also makes it sound like there's probably a better way to do this than sleep. You could just wait until there's no longer a wireguard interface present by calling ifaceExists() again?

Reply:
Added iface checks to both before a sleep. Will be curious if anyone else runs into this on Windows. The only machine I have is the one running in EC2. I added a debug statement in case anyone else is ever interested. 

Ty for the review!